File: src/main/java/org/example/kalkulationsprogramm/service/FirmeninformationService.java
package org.example.kalkulationsprogramm.service;

import com.lowagie.text.Image;
import lombok.RequiredArgsConstructor;
import org.example.kalkulationsprogramm.domain.Firmeninformation;
import org.example.kalkulationsprogramm.dto.FirmeninformationDto;
import org.example.kalkulationsprogramm.repository.FirmeninformationRepository;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.http.MediaType;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.multipart.MultipartFile;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.StandardCopyOption;
import java.util.List;
import java.util.Locale;
import java.util.Map;

@Service
@RequiredArgsConstructor
public class FirmeninformationService {

    private static final Logger log = LoggerFactory.getLogger(FirmeninformationService.class);

    /** Maximal zulaessige Logo-Dateigroesse in Bytes (2 MB). */
    static final long MAX_LOGO_BYTES = 2L * 1024 * 1024;

    /** Erlaubte MIME-Types fuer Firmenlogo (Whitelist). */
    static final List<String> ERLAUBTE_MIME_TYPES = List.of(
            MediaType.IMAGE_PNG_VALUE,
            MediaType.IMAGE_JPEG_VALUE,
            "image/webp");

    /** MIME-Type -> Standard-Dateiendung (lowercase, ohne Punkt). */
    private static final Map<String, String> MIME_ZU_ENDUNG = Map.of(
            MediaType.IMAGE_PNG_VALUE, "png",
            MediaType.IMAGE_JPEG_VALUE, "jpg",
            "image/webp", "webp");

    private final FirmeninformationRepository repository;

    /** Basis-Verzeichnis fuer Firmen-Uploads (Logo). Ueberschreibbar fuer Tests. */
    @Value("${firma.logo.upload-dir:uploads/firma/logo}")
    private String logoUploadDir;

    @Transactional(readOnly = true)
    public FirmeninformationDto getFirmeninformation() {
        Firmeninformation fi = repository.getOrCreate();
        return toDto(fi);
    }

    @Transactional
    public FirmeninformationDto speichern(FirmeninformationDto dto) {
        Firmeninformation fi = repository.getOrCreate();
        
        fi.setFirmenname(dto.getFirmenname());
        fi.setStrasse(dto.getStrasse());
        fi.setPlz(dto.getPlz());
        fi.setOrt(dto.getOrt());
        fi.setTelefon(dto.getTelefon());
        fi.setFax(dto.getFax());
        fi.setEmail(dto.getEmail());
        fi.setWebsite(dto.getWebsite());
        fi.setSteuernummer(dto.getSteuernummer());
        fi.setUstIdNr(dto.getUstIdNr());
        fi.setHandelsregister(dto.getHandelsregister());
        fi.setHandelsregisterNummer(dto.getHandelsregisterNummer());
        fi.setBankName(dto.getBankName());
        fi.setIban(dto.getIban());
        fi.setBic(dto.getBic());
        fi.setLogoDateiname(dto.getLogoDateiname());
        fi.setGeschaeftsfuehrer(dto.getGeschaeftsfuehrer());
        fi.setFusszeileText(dto.getFusszeileText());
        
        fi = repository.save(fi);
        return toDto(fi);
    }

    /**
     * Speichert eine hochgeladene Logo-Datei auf Platte und aktualisiert
     * {@link Firmeninformation#getLogoDateiname()} auf den Basenamen (ohne Pfad).
     * <p>
     * Sicherheits-Regeln:
     * <ul>
     *   <li>MIME-Whitelist: nur PNG, JPEG, WebP – sonst {@link IllegalArgumentException}</li>
     *   <li>Groessen-Limit {@value #MAX_LOGO_BYTES} Bytes</li>
     *   <li>Zielname wird <b>server-seitig</b> gesetzt ({@code logo.<ext>}),
     *       Client-Dateiname wird nie verwendet -> Pfad-Traversal ausgeschlossen</li>
     *   <li>Zielpfad wird nach Normalisierung gegen das Upload-Basis-Verzeichnis geprueft</li>
     * </ul>
     * Ein bereits existierendes Logo wird ueberschrieben (auch bei anderer Endung).
     */
    @Transactional
    public FirmeninformationDto speichereLogoDatei(MultipartFile datei) throws IOException {
        if (datei == null || datei.isEmpty()) {
            throw new IllegalArgumentException("Datei ist leer");
        }
        if (datei.getSize() > MAX_LOGO_BYTES) {
            throw new IllegalArgumentException(
                    "Datei zu groß (maximal " + (MAX_LOGO_BYTES / 1024 / 1024) + " MB)");
        }
        String contentType = datei.getContentType();
        String endung = MIME_ZU_ENDUNG.get(contentType != null ? contentType.toLowerCase(Locale.ROOT) : "");
        if (endung == null) {
            throw new IllegalArgumentException(
                    "Ungültiger Dateityp – erlaubt sind PNG, JPEG und WebP");
        }

        Path basis = Path.of(logoUploadDir).toAbsolutePath().normalize();
        Files.createDirectories(basis);

        String neuerDateiname = "logo." + endung;
        Path ziel = basis.resolve(neuerDateiname).normalize();
        if (!ziel.startsWith(basis)) {
            // kann nur bei manipuliertem logoUploadDir passieren – defensive Pruefung
            throw new IllegalStateException("Ungueltiger Zielpfad");
        }

        Firmeninformation fi = repository.getOrCreate();
        String alterDateiname = fi.getLogoDateiname();
        if (alterDateiname != null && !alterDateiname.equals(neuerDateiname)) {
            // Vorherige Datei mit abweichender Endung entfernen, damit nicht Reste liegen bleiben
            loescheLogoDateiFallsVorhanden(basis, alterDateiname);
        }

        try (var in = datei.getInputStream()) {
            Files.copy(in, ziel, StandardCopyOption.REPLACE_EXISTING);
        }

        fi.setLogoDateiname(neuerDateiname);
        fi = repository.save(fi);
        log.info("Firmenlogo gespeichert: dateiname={} groesse={}B", neuerDateiname, datei.getSize());
        return toDto(fi);
    }

    /**
     * Entfernt das Firmenlogo (Datei + Eintrag im Singleton). Idempotent:
     * ist kein Logo vorhanden, geschieht nichts.
     */
    @Transactional
    public FirmeninformationDto loescheLogoDatei() {
        Firmeninformation fi = repository.getOrCreate();
        String dateiname = fi.getLogoDateiname();
        if (dateiname != null && !dateiname.isBlank()) {
            Path basis = Path.of(logoUploadDir).toAbsolutePath().normalize();
            loescheLogoDateiFallsVorhanden(basis, dateiname);
        }
        fi.setLogoDateiname(null);
        fi = repository.save(fi);
        return toDto(fi);
    }

    /**
     * Liefert die Logo-Bytes oder {@code null}, falls kein Logo hinterlegt ist bzw.
     * die Datei auf Platte fehlt. Wird von PDF-Services benutzt, die bei
     * fehlendem Logo einfach kein Bild zeichnen.
     */
    @Transactional(readOnly = true)
    public byte[] loadLogoBytes() {
        Firmeninformation fi = repository.findFirmeninformation().orElse(null);
        if (fi == null || fi.getLogoDateiname() == null || fi.getLogoDateiname().isBlank()) {
            return null;
        }
        Path basis = Path.of(logoUploadDir).toAbsolutePath().normalize();
        Path datei = basis.resolve(fi.getLogoDateiname()).normalize();
        if (!datei.startsWith(basis) || !Files.isRegularFile(datei)) {
            return null;
        }
        try {
            return Files.readAllBytes(datei);
        } catch (IOException e) {
            log.warn("Firmenlogo konnte nicht gelesen werden: {}", datei, e);
            return null;
        }
    }

    /**
     * Liefert das Firmenlogo als iText-{@link Image} oder {@code null},
     * falls kein Logo hinterlegt ist. Aufrufer duerfen {@code null} ohne
     * Fallback akzeptieren (Design-Entscheidung: kein Software-Logo als Fallback).
     */
    public Image loadLogoImage() {
        byte[] bytes = loadLogoBytes();
        if (bytes == null) {
            return null;
        }
        try {
            return Image.getInstance(bytes);
        } catch (Exception e) {
            log.warn("Firmenlogo konnte nicht als PDF-Bild geladen werden", e);
            return null;
        }
    }

    /** Liefert die MIME-Type-Information fuer den GET-Endpoint. */
    @Transactional(readOnly = true)
    public String ermittleLogoContentType() {
        Firmeninformation fi = repository.findFirmeninformation().orElse(null);
        if (fi == null || fi.getLogoDateiname() == null) {
            return null;
        }
        String name = fi.getLogoDateiname().toLowerCase(Locale.ROOT);
        if (name.endsWith(".png")) return MediaType.IMAGE_PNG_VALUE;
        if (name.endsWith(".jpg") || name.endsWith(".jpeg")) return MediaType.IMAGE_JPEG_VALUE;
        if (name.endsWith(".webp")) return "image/webp";
        return MediaType.APPLICATION_OCTET_STREAM_VALUE;
    }

    private void loescheLogoDateiFallsVorhanden(Path basis, String dateiname) {
        if (dateiname == null || dateiname.isBlank()) {
            return;
        }
        // Dateiname kommt aus eigener DB -> trotzdem normalisiert pruefen
        String baseName = Path.of(dateiname).getFileName().toString();
        Path datei = basis.resolve(baseName).normalize();
        if (!datei.startsWith(basis)) {
            return;
        }
        try {
            Files.deleteIfExists(datei);
        } catch (IOException e) {
            log.warn("Altes Firmenlogo konnte nicht geloescht werden: {}", datei, e);
        }
    }

    private FirmeninformationDto toDto(Firmeninformation fi) {
        FirmeninformationDto dto = new FirmeninformationDto();
        dto.setId(fi.getId());
        dto.setFirmenname(fi.getFirmenname());
        dto.setStrasse(fi.getStrasse());
        dto.setPlz(fi.getPlz());
        dto.setOrt(fi.getOrt());
        dto.setTelefon(fi.getTelefon());
        dto.setFax(fi.getFax());
        dto.setEmail(fi.getEmail());
        dto.setWebsite(fi.getWebsite());
        dto.setSteuernummer(fi.getSteuernummer());
        dto.setUstIdNr(fi.getUstIdNr());
        dto.setHandelsregister(fi.getHandelsregister());
        dto.setHandelsregisterNummer(fi.getHandelsregisterNummer());
        dto.setBankName(fi.getBankName());
        dto.setIban(fi.getIban());
        dto.setBic(fi.getBic());
        dto.setLogoDateiname(fi.getLogoDateiname());
        dto.setGeschaeftsfuehrer(fi.getGeschaeftsfuehrer());
        dto.setFusszeileText(fi.getFusszeileText());
        return dto;
    }
}

File: src/main/kotlin/org/example/kalkulationsprogramm/service/FirmeninformationService.kt
package org.example.kalkulationsprogramm.service

import com.lowagie.text.Image
import org.example.kalkulationsprogramm.domain.Firmeninformation
import org.example.kalkulationsprogramm.domain.Gewerk
import org.example.kalkulationsprogramm.dto.FirmeninformationDto
import org.example.kalkulationsprogramm.repository.FirmeninformationRepository
import org.example.kalkulationsprogramm.repository.GewerkRepository
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Value
import org.springframework.http.MediaType
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import org.springframework.web.multipart.MultipartFile
import java.io.IOException
import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.StandardCopyOption
import java.util.Locale

@Service
class FirmeninformationService(
    private val repository: FirmeninformationRepository,
    private val gewerkRepository: GewerkRepository,
) {
    @Value("\${firma.logo.upload-dir:uploads/firma/logo}")
    private lateinit var logoUploadDir: String

    @Transactional(readOnly = true)
    fun getFirmeninformation(): FirmeninformationDto =
        toDto(repository.getOrCreate())

    @Transactional
    fun speichern(dto: FirmeninformationDto): FirmeninformationDto {
        var fi = repository.getOrCreate()

        fi.firmenname = dto.firmenname
        fi.strasse = dto.strasse
        fi.plz = dto.plz
        fi.ort = dto.ort
        fi.telefon = dto.telefon
        fi.fax = dto.fax
        fi.email = dto.email
        fi.website = dto.website
        fi.steuernummer = dto.steuernummer
        fi.ustIdNr = dto.ustIdNr
        fi.handelsregister = dto.handelsregister
        fi.handelsregisterNummer = dto.handelsregisterNummer
        fi.bankName = dto.bankName
        fi.iban = dto.iban
        fi.bic = dto.bic
        fi.logoDateiname = dto.logoDateiname
        fi.geschaeftsfuehrer = dto.geschaeftsfuehrer
        fi.fusszeileText = dto.fusszeileText
        fi.googleBewertungsLink = normalizeUrl(dto.googleBewertungsLink)

        fi.mahnverfahrenAktiv = dto.isMahnverfahrenAktiv
        fi.tageBisZahlungserinnerung = positivOrDefault(dto.tageBisZahlungserinnerung, 7)
        fi.tageBisErsteMahnung = positivOrDefault(dto.tageBisErsteMahnung, 14)
        fi.tageBisZweiteMahnung = positivOrDefault(dto.tageBisZweiteMahnung, 21)
        fi.mahnverfahrenNeuesZahlungszielTage = positivOrDefault(dto.mahnverfahrenNeuesZahlungszielTage, 7)

        fi.gewerk = dto.gewerkId?.let { gewerkId ->
            gewerkRepository.findById(gewerkId)
                .orElseThrow { IllegalArgumentException("Gewerk nicht gefunden: $gewerkId") }
        }
        fi.bgSatzOverride = dto.bgSatzOverride

        fi = repository.save(fi)
        return toDto(fi)
    }

    @Transactional
    @Throws(IOException::class)
    fun speichereLogoDatei(datei: MultipartFile?): FirmeninformationDto {
        if (datei == null || datei.isEmpty) {
            throw IllegalArgumentException("Datei ist leer")
        }
        if (datei.size > MAX_LOGO_BYTES) {
            throw IllegalArgumentException("Datei zu groß (maximal ${MAX_LOGO_BYTES / 1024 / 1024} MB)")
        }
        val contentType = datei.contentType
        val endung = MIME_ZU_ENDUNG[contentType?.lowercase(Locale.ROOT) ?: ""]
            ?: throw IllegalArgumentException("Ungültiger Dateityp - erlaubt sind PNG, JPEG und WebP")

        val basis = Path.of(logoUploadDir).toAbsolutePath().normalize()
        Files.createDirectories(basis)

        val neuerDateiname = "logo.$endung"
        val ziel = basis.resolve(neuerDateiname).normalize()
        if (!ziel.startsWith(basis)) {
            throw IllegalStateException("Ungueltiger Zielpfad")
        }

        var fi = repository.getOrCreate()
        val alterDateiname = fi.logoDateiname
        if (alterDateiname != null && alterDateiname != neuerDateiname) {
            loescheLogoDateiFallsVorhanden(basis, alterDateiname)
        }

        datei.inputStream.use { input ->
            Files.copy(input, ziel, StandardCopyOption.REPLACE_EXISTING)
        }

        fi.logoDateiname = neuerDateiname
        fi = repository.save(fi)
        log.info("Firmenlogo gespeichert: dateiname={} groesse={}B", neuerDateiname, datei.size)
        return toDto(fi)
    }

    @Transactional
    fun loescheLogoDatei(): FirmeninformationDto {
        var fi = repository.getOrCreate()
        val dateiname = fi.logoDateiname
        if (!dateiname.isNullOrBlank()) {
            val basis = Path.of(logoUploadDir).toAbsolutePath().normalize()
            loescheLogoDateiFallsVorhanden(basis, dateiname)
        }
        fi.logoDateiname = null
        fi = repository.save(fi)
        return toDto(fi)
    }

    @Transactional(readOnly = true)
    fun loadLogoBytes(): ByteArray? {
        val fi = repository.findFirmeninformation().orElse(null)
        if (fi == null || fi.logoDateiname.isNullOrBlank()) {
            return null
        }
        val basis = Path.of(logoUploadDir).toAbsolutePath().normalize()
        val datei = basis.resolve(fi.logoDateiname).normalize()
        if (!datei.startsWith(basis) || !Files.isRegularFile(datei)) {
            return null
        }
        return try {
            Files.readAllBytes(datei)
        } catch (e: IOException) {
            log.warn("Firmenlogo konnte nicht gelesen werden: {}", datei, e)
            null
        }
    }

    @Transactional(readOnly = true)
    fun loadLogoImage(): Image? {
        val bytes = loadLogoBytes() ?: return null
        return try {
            Image.getInstance(bytes)
        } catch (e: Exception) {
            log.warn("Firmenlogo konnte nicht als PDF-Bild geladen werden", e)
            null
        }
    }

    @Transactional(readOnly = true)
    fun ermittleLogoContentType(): String? {
        val fi = repository.findFirmeninformation().orElse(null)
        val name = fi?.logoDateiname?.lowercase(Locale.ROOT) ?: return null
        return when {
            name.endsWith(".png") -> MediaType.IMAGE_PNG_VALUE
            name.endsWith(".jpg") || name.endsWith(".jpeg") -> MediaType.IMAGE_JPEG_VALUE
            name.endsWith(".webp") -> "image/webp"
            else -> MediaType.APPLICATION_OCTET_STREAM_VALUE
        }
    }

    private fun loescheLogoDateiFallsVorhanden(basis: Path, dateiname: String?) {
        if (dateiname.isNullOrBlank()) {
            return
        }
        val baseName = Path.of(dateiname).fileName.toString()
        val datei = basis.resolve(baseName).normalize()
        if (!datei.startsWith(basis)) {
            return
        }
        try {
            Files.deleteIfExists(datei)
        } catch (e: IOException) {
            log.warn("Altes Firmenlogo konnte nicht geloescht werden: {}", datei, e)
        }
    }

    private fun toDto(fi: Firmeninformation): FirmeninformationDto {
        val dto = FirmeninformationDto()
        dto.id = fi.id
        dto.firmenname = fi.firmenname
        dto.strasse = fi.strasse
        dto.plz = fi.plz
        dto.ort = fi.ort
        dto.telefon = fi.telefon
        dto.fax = fi.fax
        dto.email = fi.email
        dto.website = fi.website
        dto.steuernummer = fi.steuernummer
        dto.ustIdNr = fi.ustIdNr
        dto.handelsregister = fi.handelsregister
        dto.handelsregisterNummer = fi.handelsregisterNummer
        dto.bankName = fi.bankName
        dto.iban = fi.iban
        dto.bic = fi.bic
        dto.logoDateiname = fi.logoDateiname
        dto.geschaeftsfuehrer = fi.geschaeftsfuehrer
        dto.fusszeileText = fi.fusszeileText
        dto.googleBewertungsLink = fi.googleBewertungsLink
        dto.isMahnverfahrenAktiv = fi.isMahnverfahrenAktiv()
        dto.tageBisZahlungserinnerung = fi.tageBisZahlungserinnerung
        dto.tageBisErsteMahnung = fi.tageBisErsteMahnung
        dto.tageBisZweiteMahnung = fi.tageBisZweiteMahnung
        dto.mahnverfahrenNeuesZahlungszielTage = fi.mahnverfahrenNeuesZahlungszielTage
        val gewerk: Gewerk? = fi.gewerk
        if (gewerk != null) {
            dto.gewerkId = gewerk.id
            dto.gewerkName = gewerk.name
            dto.bgName = gewerk.bgName
            dto.bgSatzVorschlag = gewerk.bgSatzProzent
        }
        dto.bgSatzOverride = fi.bgSatzOverride
        dto.bgSatzEffektiv = fi.bgSatzOverride ?: gewerk?.bgSatzProzent
        return dto
    }

    companion object {
        private val log = LoggerFactory.getLogger(FirmeninformationService::class.java)
        const val MAX_LOGO_BYTES: Long = 2L * 1024 * 1024

        val ERLAUBTE_MIME_TYPES: List<String> = listOf(
            MediaType.IMAGE_PNG_VALUE,
            MediaType.IMAGE_JPEG_VALUE,
            "image/webp",
        )

        private val MIME_ZU_ENDUNG: Map<String, String> = mapOf(
            MediaType.IMAGE_PNG_VALUE to "png",
            MediaType.IMAGE_JPEG_VALUE to "jpg",
            "image/webp" to "webp",
        )

        private fun positivOrDefault(wert: Int, fallback: Int): Int =
            if (wert > 0) wert else fallback

        private fun normalizeUrl(url: String?): String? {
            val trimmed = url?.trim()
            return if (trimmed.isNullOrEmpty()) null else trimmed
        }
    }
}
